### PR TITLE
ci: Enable testing on M1 macOS

### PR DESCRIPTION
**Pull request by QuLogic**
## PR summary

Just added in beta today:
https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

Need to see which of the optional dependencies are okay with this.

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by QuLogic** [comment]
Minimum Python available there is 3.10 now, so bumped that up. It looks like `pycairo` and `PyGObject` don't have wheels. PyQt5 and wxPython4 have wheels, but I haven't tried allowing the others that were blocked on macOS yet.

We do seem to be getting a lot more failures here than we did when trying Cirrus in #24597.

**Comment by QuLogic** [comment]
I've de-conditionalized all optional GUI dependencies so I can verify whether they should be skipped or not.

**Comment by QuLogic** [comment]
The results are the following:
| System                     | PyGObject3         | PyGObject4         | PyQt5              | PySide2            | PyQt6              | PySide6            | wxPython           |
| -------------------------- | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ |
| Py3.9 Ubuntu 20.04 MinVer  | :heavy_check_mark: | :x:                | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Py3.9 Ubuntu 20.04 Extras  | :heavy_check_mark: | :x:                | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Py3.10 Ubuntu 20.04 Extras | :heavy_check_mark: | :x:                | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Py3.11 Ubuntu 22.04 Extras | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Py3.12 Ubuntu 22.04        | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :bangbang:         | :heavy_check_mark: | :heavy_check_mark: | :x:                |
| macOS 12 amd64             | :x:                | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| macOS 14 arm64             | :x:                | :ferris_wheel:     | :heavy_check_mark: | :x:                | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

Most are available everywhere.

The PyGObject 3/4 split doesn't have any effect, as the dependencies are the same either way (the only difference is the native dependency.) Only PyGObject 4 on macOS 14 arm64 is missing wheels, but appears to be available regardless after building them.

PySide2 on Python 3.12 installs a wheel, but crashes. I believe PySide2 is no longer supported so they didn't bother fixing it for Python 3.12, and because it's some kind of universal thing, it still installed. On the other hand the M1 was new enough that there were no wheels for it.

**Comment by QuLogic** [comment]
I rebased #24597 for comparison, and it doesn't seem to fail nearly as many tests. I can't see any tolerance issues at all. For a reason I don't understand, `xfail` appears to be broken on both.

**Comment by QuLogic** [comment]
Ah, looking at the [pytest 8 changelog](https://docs.pytest.org/en/8.0.x/changelog.html#improvements), xfails are now fully reported with backtraces. This seems far too noisy, and we may have to disable that.

**Comment by tacaswell** [comment]
```
(mpl-dev) tcaswell@seven matplotlib %  python -c 'from matplotlib.backends._macosx import print_screen_info; print_screen_info()'                                                                                                                                                                                      macos-m1
mainScreen: Acer B286HK (3840.000000x2160.000000)+0.000000+0.000000 @8
deepestScreen: Acer B286HK (3840.000000x2160.000000)+0.000000+0.000000 @8
Available screens: 1
    0: Acer B286HK (3840.000000x2160.000000)+0.000000+0.000000 @8
(mpl-dev) tcaswell@seven matplotlib %  python -c 'from matplotlib.backends._macosx import print_screen_info; print_screen_info()'                                                                                                                                                                                      macos-m1
mainScreen: (unknown)
deepestScreen: (unknown)
Available screens: 0
(mpl-dev) tcaswell@seven matplotlib %  python -c 'from matplotlib.backends._macosx import print_screen_info; print_screen_info()'                                                                                                                                                                                      macos-m1
mainScreen: (unknown)
deepestScreen: (unknown)
Available screens: 0
```

That is the result for me running over ssh with:

 - account logged in and a screen connected (but locked)
 - account logged out and a screen connected
 - account logged out and no screen connected

The account I was logged in via and the account I was ssh'd to were the same.  My guess is that being logged in as a different user on the screen would still result in no screens being reported via ssh (but I have not tested that if we care I can).

We should definitely adapt this and use it the same way we use the "do we have a screen" check on linux to auto fallback to Agg.

**Comment by QuLogic** [comment]
OK, I removed the debugging, and pushed the screen check to #27761.

**Comment by QuLogic** [comment]
OK, it's interesting that PyQt5 started failing even though that was always installed and it seems to still be the same version. I enabled PySide2, PyQt6, and PySide6, which could maybe conflict, but they don't seem to have installed any different packages.

**Comment by QuLogic** [comment]
I pushed https://github.com/matplotlib/matplotlib/pull/27723/commits/17dda1d040a9a5806fecb3dd97f09dbebe586249 to modify the cross Qt import test to be more resiliant and actually test what we are interested in. I also pinged the meson-python issue to see if they can make a new release https://github.com/mesonbuild/meson-python/issues/524#issuecomment-2005876236

**Comment by QuLogic** [comment]
There seems to be a bug with `subprocess` on the M1 with Python 3.10. It is documented as:

> The timeout argument is passed to [Popen.communicate()](https://docs.python.org/3.10/library/subprocess.html#subprocess.Popen.communicate). If the timeout expires, the child process will be killed and waited for. The [TimeoutExpired](https://docs.python.org/3.10/library/subprocess.html#subprocess.TimeoutExpired) exception will be re-raised after the child process has terminated.

However, the child process doesn't seem to have been killed, as the `Popen.__del__` warning leaks into some other test instead.

Bumping M1 to Python 3.12 appears to have fixed it, though some of the Qt 6 tests are failing the sigint tests. Since they weren't tested before, I might have to mark them as xfail for now.

**Comment by QuLogic** [comment]
So I added a check for an exception in the signal handler, printed it out, then re-raised it: https://github.com/matplotlib/matplotlib/pull/27723/commits/c1afbac74d6b64ced2954cb3d68b1ed79c0af6bd

But that magically makes the whole thing work, even though no exception is printed or re-raised, and coverage says those lines were not run. And this has worked twice now (with it still broken for a run in between where I removed the last patch.)

**Comment by QuLogic** [comment]
So I rebased and decided to try leaving out the exception change, and it's all working again somehow...

We also decided to add Ghostscript and Inkscape to see how that goes.

**Comment by QuLogic** [comment]
We decided to xfail the `test_other_signal_before_sigint` test, because sometimes seems fixed, and sometimes doesn't, regardless of the commit above, and @ksunden reproduced _once_, but not much luck after that.

Also, contrary to our call @ksunden/@tacaswell, I decided to keep the native arm64 build here. In fact, the cross-compiled one is brokenly-named as well, so we need to fix it one way or the other, and might as well do it now.

**Comment by QuLogic** [comment]
Because this fixes the wheel names, this is important to get in to 3.9 before the rc.

File: lib/matplotlib/tests/test_path.py [hunk 1]
```diff
@@ -281,7 +282,8 @@ def test_marker_paths_pdf():
 
 
 @image_comparison(['nan_path'], style='default', remove_text=True,
-                  extensions=['pdf', 'svg', 'eps', 'png'])
+                  extensions=['pdf', 'svg', 'eps', 'png'],
+                  tol=0.009 if platform.machine() else 0)
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_path.py:286]:
```suggestion
                  tol=0.009 if platform.machine() == 'arm64' else 0)
```

**Comment by tacaswell** [comment]
I only half understand the comment in the svg backend.

File: lib/matplotlib/tests/test_path.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/tests/test_path.py:286]: Oops, thought I caught all these copy-paste typos.

**Comment by QuLogic** [comment]
I made some edits to those comments.

**Review by tacaswell — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Pull request merged**